### ISSUE TITLE
OneHotEncoding should retain metadata

### ISSUE THREAD

**Issue by rofinn** [invenia/FeatureTransforms.jl#57]
The current `OneHotEncoding` transform seems to always just return a NxP binary matrix despite knowing what the categories are and potentially being passed a type that would allow you to retain that info.

For example, if I have `OneHotEncoding(Hour(0):Hour(1):Hour(23))` and then I pass a dataframe with an `HoD` column, I could easily see that transform returning a new dataframe with a column name for each hour. Similarly, if we took a function argument, we could do something like:

```julia
FeatureTransforms.ohe(Hour, data; dims=:time)
```

This would return a new `KeyedArray` or `AxisArray` that retains the category information for the `p` dimension, and also uses the `dims` argument to mean use dimension keys for `KeyedArray`s or `AxisArrays`.

**Comment by glennmoy** [comment]
Can you explain what you had in mind for the function argument? I'm not sure I see how it translates to keeping the category information? Is it that the function determines what the category labels are?

**Comment by rofinn** [comment]
>  Is it that the function determines what the category labels are?

Yes, so if I have an axis with datetimes I might use `hour` to lazily generate the categories. For `AxisArray`s and `KeyedArray`s we could then return n (`:time`/`DateTime`) x p (`:category`/`Hour`).
